README.rst
20min
=====

.. image:: https://gitlab.com/grauwoelfchen/20min/badges/trunk/pipeline.svg
   :target: https://gitlab.com/grauwoelfchen/20min/commits/trunk

.. image:: https://gitlab.com/grauwoelfchen/20min/badges/trunk/coverage.svg
   :target: https://gitlab.com/grauwoelfchen/20min/commits/trunk

.. image:: https://img.shields.io/crates/v/twenty-minutes?label=crates&style=flat
   :target: https://crates.io/crates/twenty-minutes

.. image:: https://docs.rs/twenty-minutes/badge.svg
   :target: https://docs.rs/crate/twenty-minutes

|

.. raw:: html

   <del>News von jetzt!</del>

``20min`` (twenty-minutes) is a command line working timer.

Install
-------

.. code:: zsh

   % cargo install twenty-minutes


Usage
-----

.. code:: zsh

   # 900 secs work and 300 secs rest (same as default)
   % 20min -w 15 -r 5
       Working [==============================================================] 100%
      Finished work [unoptimized + progressbar] targes(+s) in 900 secs
       Resting [===================================================-----------] 80%


.. code:: zsh

   # see 20min --help for the details
   % 20min
   % 20min 15,5
   % 20min 15 5
   % 20min --work 15 --rest 5
   % 20min -w 15


Build
-----

Check ``make help``

.. code:: zsh

   # debug build
   % make build


Test
-----

.. code:: zsh

   % make test
   % make test:coverage


License
-------


.. code:: text

   20min
   Copyright 2018-2022 Yasuhiro Яша Asaka

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


CHANGELOG
Sun 17 Jul 2022 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	New release

	* Create v0.0.3

	Update edition to 2021

	Refactor integration tests

Sun 02 May 2021 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Release v0.0.1 and v0.0.2
	* Separate lib as twenty_minutes
	* Enable pages build

Sat 01 May 2021 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Add runner-% in Makefile
	* Remove .tool directory

Fri 25 Jul 2020 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Set rust version as stable
	* Fix jobs on CI
	* Set package name as "cli-20min" for now (can't start it with a number)
	* Fix coverage report (Update kcov from v34 to v38)

Thu 18 Jul 2019 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Update dependencies (structopt and cargo-husky)
	* Rename tools directory as .tool, tests as test

Sun 18 Nov 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Disable pages stage on CI
	* Add commit hooks using cargo-husky
	* Change indent spaces as 4 (replace 4 spaces as 2 space in the editor)

Thu 08 Nov 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Add make coverage targets
	* Add make vet:all and make vet
	* Add abort on panic for release build

Tue 11 Sep 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Update gitlab-runner (gitlab-ci-multi-runner -> gitlab-runner)

Wed 29 Aug 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Update rustfmt-preview and clippy
	* Add `--work/--rest` options (see 20min --help for details)

Fri 13 Apr 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Update docker images on CI (Gentoo Linux)

Thu 06 Apr 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Update images on CI to grauwoelfchen/rust (gentoo)
	* Fix print_literal (lint)

Wed 05 Apr 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Make code more rust like one ;)
	* Add `make lint` using clippy (.clippy.toml)
	* Add `make help`

Thu 30 Mar 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Add gitlab pages (documents)
	* Add documents for main crate and functions
	* Add .rustfmt.toml for style formatting

Mon 19 Mar 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Add format target
	* Move scripts into tools

Mon 05 Mar 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Update kcov (local/ci) installation to fix coverage
	* Add a runner script for coverage

Tue 09 Jan 2018 Yasuhiro Яша Asaka <yasuhiro.asaka@grauwoelfchen.net>
	* Initial commit ;)


test/args.rs
#[cfg(test)]
mod test {
    use crate::run;

    #[test]
    fn test_run_with_invalid_args_and_comma() {
        let output = run(&["foo,bar"]).expect("command returns output");
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));
    }

    #[test]
    fn test_run_with_invalid_args_no_comma() {
        let output = run(&["foo"]).expect("command returns output");
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));
    }

    #[test]
    fn test_run_with_valid_args_and_comma() {
        // work: 0.6 secs, rest: 0.6 secs
        let output = run(&["0.01,0.01"]).expect("command returns output");
        assert_eq!(String::from_utf8_lossy(&output.stderr), "");
    }

    #[test]
    fn test_run_with_valid_args_no_comma() {
        // work: 0.6 secs, rest: 0.6 secs
        let output = run(&["0.01", "0.01"]).expect("command returns output");
        assert_eq!(String::from_utf8_lossy(&output.stderr), "");
    }
}


test/opts.rs
#[cfg(test)]
mod test {
    use crate::run;

    #[test]
    fn test_run_with_invalid_opts() {
        let output = run(&["--work", "foo"]).unwrap();
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));

        let output = run(&["--rest", "foo"]).unwrap();
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));

        let output = run(&["--work", "3", "--rest", "foo"]).unwrap();
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));

        let output = run(&["--work", "foo", "--rest", "3"]).unwrap();
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));

        let output = run(&["--work", "foo", "--rest", "bar"]).unwrap();
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));
    }

    #[test]
    fn test_run_with_unknown_opts() {
        let output = run(&["--no-more-work"]).unwrap();
        assert!(String::from_utf8_lossy(&output.stderr).contains("panicked"));
    }

    #[test]
    fn test_run_with_valid_opts() {
        let output = run(&["--work", "0.01", "--rest", "0.01"]).unwrap();
        assert_eq!(String::from_utf8_lossy(&output.stderr), "");

        let output = run(&["-w", "0.01", "-r", "0.01"]).unwrap();
        assert_eq!(String::from_utf8_lossy(&output.stderr), "");
    }
}


test/test.rs
mod args;
mod opts;

use std::env;
use std::process::{Command, Output};

const DEFAULT_CARGO_TARGET_DIR: &str = "./target";

fn run(args: &[&str]) -> Result<Output, std::io::Error> {
    let target_dir = env::var("CARGO_TARGET_DIR")
        .unwrap_or_else(|_| DEFAULT_CARGO_TARGET_DIR.to_string());
    let program = format!("{}/debug/20min", target_dir);

    Command::new(&program).args(args).output()
}


src/lib.rs
use std::str::FromStr;

pub mod config;
pub mod ticker;

/// Create new pair using `Option`.
///
/// This is a private utility function to parse command line args from string
/// into pair.
///
/// # Errors
///
/// This function will not return any errors, otherwise returns None.
///
/// # Examples
///
/// ```
/// assert_eq!(None, parse_pair::<i32>("", ','));
/// assert_eq!(Some("15", "5"), parse_pair::<i32>("15,5", ','));
/// ```
fn parse_pair<T: FromStr>(s: &str, separator: char) -> Option<(T, T)> {
    match s.find(separator) {
        None => None,
        Some(index) => {
            match (T::from_str(&s[..index]), T::from_str(&s[index + 1..])) {
                (Ok(l), Ok(r)) => Some((l, r)),
                _ => None,
            }
        }
    }
}

/// Parses an arg into Vec.
///
/// # Errors
///
/// This private function will not return any errors, otherwise returns None.
///
/// # Examples
///
/// ```
/// assert_eq!(
///     Some(vec!["10".to_string(), "".to_string()]),
///     split_arg("10,")
/// );
/// ```
pub fn split_arg(s: &str) -> Option<Vec<String>> {
    parse_pair::<String>(s, ',').map(|(w, r)| vec![w, r])
}

/// Outputs the formatted result.
pub fn report(duration: u64) {
    let prefix = "Finished";
    let prefix_width = 12;
    let info = "work [unoptimized + progressbar]";

    println!(
        "{prefix:>prefix_width$} {info} target(s) in {duration} secs\r",
        prefix = prefix,
        prefix_width = prefix_width,
        info = info,
        duration = duration
    );
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_parse_pair() {
        assert_eq!(parse_pair::<i32>("", ','), None);
        assert_eq!(parse_pair::<i32>("10,", ','), None);
        assert_eq!(parse_pair::<i32>(",10", ','), None);
        assert_eq!(parse_pair::<i32>("10,20", ','), Some((10, 20)));
        assert_eq!(parse_pair::<i32>("10,20xy", ','), None);
        assert_eq!(parse_pair::<f64>("0.5x", 'x'), None);
        assert_eq!(parse_pair::<f64>("0.5x1.5", 'x'), Some((0.5, 1.5)));
        assert_eq!(
            parse_pair::<String>("0,5", ','),
            Some(("0".to_string(), "5".to_string()))
        );
        assert_eq!(
            parse_pair::<String>("0.5,7.8", ','),
            Some(("0.5".to_string(), "7.8".to_string()))
        );
    }

    #[test]
    fn test_split_arg() {
        assert_eq!(split_arg(""), None);
        assert_eq!(
            split_arg("10,"),
            Some(vec!["10".to_string(), "".to_string()])
        );
        assert_eq!(
            split_arg(",10"),
            Some(vec!["".to_string(), "10".to_string()])
        );
        assert_eq!(
            split_arg("10,20"),
            Some(vec!["10".to_string(), "20".to_string()])
        );
        assert_eq!(
            split_arg("10,20xy"),
            Some(vec!["10".to_string(), "20xy".to_string()])
        );
        assert_eq!(
            split_arg("0.5,x"),
            Some(vec!["0.5".to_string(), "x".to_string()])
        );

        assert_eq!(split_arg("0.5x"), None);
        assert_eq!(split_arg("0.5x1.5"), None);
    }
}


src/config.rs
use std::str::FromStr;

pub struct Config {
    pub params: Vec<u64>,
}

impl Config {
    const PARSE_ERROR: &'static str = "Error: Unknown argument";

    pub fn new(args: Vec<String>) -> Self {
        let mut p: Vec<u64> = Vec::new();
        for arg in args {
            // String -> u64
            let v: u64 = if arg.contains('.') {
                let f = f64::from_str(&arg).expect(Config::PARSE_ERROR);
                (f * 60.0).round() as u64
            } else {
                u64::from_str(&arg).expect(Config::PARSE_ERROR) * 60
            };
            p.push(v);
        }
        Config { params: p }
    }

    pub fn to_tuple(&self) -> (u64, u64) {
        // defaults (seconds)
        let work_t = 15 * 60;
        let rest_t = 5 * 60;

        let vec = &self.params;
        match vec.len() {
            0 => (work_t, rest_t),
            1 => (vec[0], rest_t),
            2 => (vec[0], vec[1]),
            _ => panic!("Error: Too many values"),
        }
    }
}

#[cfg(test)]
mod config_test {
    use std::panic;
    use super::*;

    #[test]
    fn test_config_to_tuple() {
        let c = Config { params: vec![] };
        assert_eq!((900, 300), c.to_tuple());
        assert_eq!(c.to_tuple(), c.to_tuple());

        let c = Config { params: vec![10] };
        assert_eq!((10, 300), c.to_tuple());

        let c = Config {
            params: vec![20, 30],
        };
        assert_eq!((20, 30), c.to_tuple());

        let c = Config {
            params: vec![305, 45],
        };
        assert_eq!((305, 45), c.to_tuple());

        let c = Config {
            params: vec![30, 40, 50],
        };
        let res = panic::catch_unwind(|| c.to_tuple());
        assert!(res.is_err());
    }
}


src/ticker.rs
//! A collection as module which have functions that handle the progress.
use std::io::{self, Write};
use std::{thread, time};

// TODO: termcolor support
// https://github.com/rust-lang/cargo/blob/master/src/cargo/core/shell.rs

/// Prints current progress as bar
///
/// # Examples
///
/// ```
/// assert_eq!("", progress(10, 100, "Working"));
/// ```
pub fn progress(count: u64, total: u64, prefix_text: &str) {
    let bar_len = 62; // 80 - (12 + 6)
    let text_width = 12;
    let filled_len = (bar_len * count) / total;
    let value = (100 * count) / total;
    let mut bar = "=".repeat(filled_len as usize);
    bar.push_str(&"-".repeat((bar_len - filled_len) as usize));

    print!(
        "{text:>text_width$} [{bar}] {value}%\r",
        text = &prefix_text,
        text_width = text_width,
        bar = bar,
        value = value
    );
    io::stdout().flush().unwrap();
}

pub fn tick(total: u64, prefix_text: &str) {
    if total > 0 {
        progress(0, total, prefix_text);
        for i in 1..=total {
            thread::sleep(time::Duration::new(1, 0));
            progress(i, total, prefix_text);
        }
        println!();
    }
}


src/main.rs
//! A command line working timer `20min` (twenty-minutes).
//!
//! # Examples
//!
//! ```zsh
//! % 20min 15,5
//! % 20min 0.5 5
//! ```
extern crate structopt;

use std::process::exit;

use structopt::StructOpt;

use twenty_minutes::{config::Config, ticker::tick, split_arg, report};

/// Command line options.
#[derive(Debug, StructOpt)]
#[structopt(
    name = "20min",
    about = "\nA command line working timer. Run with arguments `15,5` \
             (minutes) or OPTIONS \nlike `--work 15 --rest 15`."
)]
struct Opts {
    /// Working time
    #[structopt(short = "w", long = "work", default_value = "15")]
    work: String,

    /// Time of rest
    #[structopt(short = "r", long = "rest", default_value = "5")]
    rest: String,
}

fn main() {
    let mut args: Vec<String> = std::env::args().skip(1).collect();

    // detect argument type ("20,3" or "--work 5 --rest 0.1")
    let mut use_opts: bool = false;
    let o = [
        "-w",
        "--work",
        "-r",
        "--rest",
        "-h",
        "--help",
        "-V",
        "--version",
    ];
    for (_, opt) in o.iter().enumerate() {
        if args.contains(&(opt.to_string())) {
            use_opts = true;
            break;
        }
    }

    if use_opts {
        let opts = Opts::from_args();
        args = vec![opts.work, opts.rest];
    }

    if args.len() == 1 {
        // handle single string input like: "15,5"
        if let Some(v) = split_arg(&args[0]) {
            args = v;
        }
    }

    let c = Config::new(args);
    if c.params.len() > 2 {
        eprintln!("Usage: 20min n (working time) n (time for rest)");
        exit(1);
    }

    let (work_t, rest_t) = c.to_tuple();

    tick(work_t, "Working");
    report(work_t);

    tick(rest_t, "Resting");
    report(rest_t);
}
